src/hypothesis/extra/datetime.py

```diff
@@ -23,62 +32,185 @@
+        self.min_dt = ensure_datetime_bound(min_datetime, 'min_datetime')
+        self.max_dt = ensure_datetime_bound(max_datetime, 'max_datetime')
+
+        if (self.min_dt.tzinfo is None) != (self.max_dt.tzinfo is None):
+            # Fix the "only supplied one bound" case, or raise InvalidArgument
+            if self.min_dt.tzinfo is None and self.min_dt == dt.datetime.min:
+                self.min_dt = pytz.utc.localize(dt.datetime.min)
+            elif self.max_dt.tzinfo is None and self.max_dt == dt.datetime.max:
+                self.max_dt = pytz.utc.localize(dt.datetime.max)
+            else:
+                raise InvalidArgument((
+                    'min_datetime=%s and max_datetime=%s must not mix naive '
+                    'and tz-aware values.  See the documentation for how each '
+                    'of these modes work.') % (self.min_dt, self.max_dt))
+        if allow_naive and self.min_dt.tzinfo is not None:
+            raise InvalidArgument(
+                'Cannot create naive datetimes between tz-aware bounds as '
+                'they are not comparable; do not set allow_naive=True.')
+        check_valid_interval(
+            self.min_dt, self.max_dt, 'min_datetime', 'max_datetime')
 
     def do_draw(self, data):
         while True:
             try:
-                result = dt.datetime(
-                    year=cu.centered_integer_range(
-                        data, self.min_year, self.max_year, 2000
-                    ),
-                    month=cu.integer_range(data, 1, 12),
-                    day=cu.integer_range(data, 1, 31),
-                    hour=cu.integer_range(data, 0, 24),
-                    minute=cu.integer_range(data, 0, 59),
-                    second=cu.integer_range(data, 0, 59),
-                    microsecond=cu.integer_range(data, 0, 999999)
-                )
-                if (
-                    not self.allow_naive or
-                    (self.timezones and cu.boolean(data))
-                ):
-                    result = cu.choice(data, self.timezones).localize(result)
-                return result
-
+                # Constuct a datetime within bounds, ignoring tzinfo for now
+                result = dt.datetime.min
+                constrained = True
+                for name in ('year', 'month', 'day',
+                             'hour', 'minute', 'second', 'microsecond'):
+                    low = getattr(dt.datetime.min, name)
+                    high = getattr(dt.datetime.max, name)
+                    if constrained:
+                        low = max(low, getattr(self.min_dt, name))
+                        high = min([high, getattr(self.max_dt, name)])
+                    if name == 'year':
+                        val = cu.centered_integer_range(data, low, high, 2000)
+                    else:
+                        val = cu.integer_range(data, low, high)
+                    result = result.replace(**{name: val})
```

Comment:
nit: I think it's probably fine, and almost certainly lost in the noise of general Hypothesis performance, but repeatedly calling replace in a loop seems like a bad idea in that it has to do a fair bit of construction and validation in there. I'd probably prefer assigning these into a dict and then doing a big update at the end.